### ISSUE TITLE
nixos-unstable is blocked because of failing luks-format1 test

### ISSUE THREAD

**Issue by doronbehar** [NixOS/nixpkgs#96479]
## Issue description

https://hydra.nixos.org/job/nixos/trunk-combined/nixos.tests.installer.luksroot.x86_64-linux
https://hydra.nixos.org/build/126074627
https://hydra.nixos.org/build/126074627/nixlog/1

### Steps to reproduce

```
nix-build nixos/tests/installer.nix -A luksroot-format1
```

**Comment by flokli** [comment]
Hydra shows a kernel panic.

**Comment by doronbehar** [comment]
When I tested it locally it failed with exactly the same log tail.

**Comment by B4dM4n** [comment]
Increasing the kernel loglevel to 8 reveals the following log:

```log
boot-after-install: [    0.530834] Freeing unused kernel image memory: 1976K
boot-after-install: [    0.531483] Run /init as init process
boot-after-install: [    0.545577] stage-1-init: loading module virtio_balloon...
boot-after-install: [    0.550698] stage-1-init: loading module virtio_console...
boot-after-install: [    0.554044] stage-1-init: loading module virtio_rng...
boot-after-install: [    0.557359] stage-1-init: loading module dm_mod...
boot-after-install: [    0.563790] device-mapper: ioctl: 4.41.0-ioctl (2019-09-16) initialised: dm-devel@redhat.com
boot-after-install: [    0.565252] stage-1-init: running udev...
boot-after-install: [    0.570045] stage-1-init: Starting version 245
boot-after-install: performing optical character recognition
boot-after-install: sending monitor command: screendump /build/tmp1o6ne8yu
--
boot-after-install: [    0.910582] cdrom: Uniform CD-ROM driver Revision: 3.20
boot-after-install: [    0.912114] sr 1:0:0:0: Attached scsi CD-ROM sr0
boot-after-install: [    1.180542] stage-1-init: starting device mapper and LVM...
boot-after-install: [    1.347791] stage-1-init: hwclock: settimeofday: Invalid argument
boot-after-install: [    1.348704] clocksource: Switched to clocksource acpi_pm
boot-after-install: [    2.397577] SELinux: unrecognized netlink message: protocol=0 nlmsg_type=108 sclass=netlink_route_socket pid=108 comm=systemd-udevd
--
boot-after-install: [   20.326962] SELinux: unrecognized netlink message: protocol=0 nlmsg_type=108 sclass=netlink_route_socket pid=108 comm=systemd-udevd
boot-after-install: [   21.406098] SELinux: unrecognized netlink message: protocol=0 nlmsg_type=108 sclass=netlink_route_socket pid=108 comm=systemd-udevd
boot-after-install: [   22.424927] stage-1-init: waiting for device /dev/mapper/cryptroot to appear.......................
boot-after-install: [   22.426451] stage-1-init: Timed out waiting for device /dev/mapper/cryptroot, trying to mount anyway.
boot-after-install: [   22.445943] SELinux: unrecognized netlink message: protocol=0 nlmsg_type=108 sclass=netlink_route_socket pid=108 comm=systemd-udevd
boot-after-install: [   22.557292] stage-1-init: mounting /dev/mapper/cryptroot on /...
boot-after-install: [   22.607374] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
boot-after-install: [   22.608703] CPU: 0 PID: 1 Comm: init Not tainted 5.4.61 #1-NixOS
```

This shows that `stage-1-init` can't mount `/dev/mapper/cryptroot`, because the `dm_crypt` module is not loaded.

Comparing the the log section `Create the NixOS configuration` of the currently failing test run (https://hydra.nixos.org/build/126109626) with the last successful one (https://hydra.nixos.org/build/125651314) shows that the `hardware-configuration.nix` is missing the
```nix
boot.initrd.luks.devices."cryptroot".device = "...";
```

line.

So `nixos-install` failed to detect the encrypted device and this probably happened between 70cfd9d25e1d4f5a40f5d0a518f0749635792667 .. 8e6d8fe9f37a4565aa3c9bfe938003118c96e1f0

**Comment by flokli** [comment]
I am currently bisecting this, will post here once I have some results.

**Comment by flokli** [comment]
`git bisect` said this:

```
72c8ed038946850d345db13f30e259a368d7c3c5 is the first bad commit
commit 72c8ed038946850d345db13f30e259a368d7c3c5
Author: Florian Klink <flokli@flokli.de>
Date:   Mon Aug 19 02:08:46 2019 +0200

    systemd: build with cryptsetup and cryptsetup-generators
    
    There's a circular dependency to systemd via cryptsetup and lvm2
    (systemd -> cryptsetup -> lvm2 -> udev=systemd).
    
    However, cryptsetup only really needs the devmapper component shipped
    with lvm2. So build `pkgs.cryptsetup` with a lvm2 that doesn't come with
    udev.

 nixos/modules/system/boot/systemd.nix      | 2 +-
 pkgs/os-specific/linux/systemd/default.nix | 4 +++-
 pkgs/top-level/all-packages.nix            | 6 +++++-
 3 files changed, 9 insertions(+), 3 deletions(-)
```

```
git bisect log
git bisect start
# bad: [86fa61033a96d3b58838bb399cc16514a0b5c71e] Merge pull request #96008 from baloo/baloo/bugfixes/make-modules-closure_firmware
git bisect bad 86fa61033a96d3b58838bb399cc16514a0b5c71e
# good: [c59ea8b8a0e7f927e7291c14ea6cd1bd3a16ff38] Merge pull request #95839 from WilliButz/update/grafana-7.1.3
git bisect good c59ea8b8a0e7f927e7291c14ea6cd1bd3a16ff38
# skip: [9eeff74821ffb6be45cccb0a955c942a012437db] Merge pull request #96182 from oxalica/rust-analyzer
git bisect skip 9eeff74821ffb6be45cccb0a955c942a012437db
# skip: [a93c4f512953395073748d8d79f10f02f4ee45f7] python3Packages.matplotlib: fix deps
git bisect skip a93c4f512953395073748d8d79f10f02f4ee45f7
# good: [609cdfae9aa71ddc74e9f01ef492a19ba5482a05] freeradius: 3.0.20 -> 3.0.21
git bisect good 609cdfae9aa71ddc74e9f01ef492a19ba5482a05
# good: [f20f270e9a363f0882108159d95396dcf59fb5b2] nextdns: 1.7.0 -> 1.7.1
git bisect good f20f270e9a363f0882108159d95396dcf59fb5b2
# bad: [2f5277d92eca0b19096255a67c65bd4864ae54c1] nix-build-uncached: remove the nix wrap (#96357)
git bisect bad 2f5277d92eca0b19096255a67c65bd4864ae54c1
# bad: [24c18f13138360136b2a53b33b01c82e36490572] linux: 5.9-rc1 -> 5.9-rc2
git bisect bad 24c18f13138360136b2a53b33b01c82e36490572
# good: [03964e52b40738bd3066c78f2084ee5f9016c96e] Merge pull request #89546 from andehen/init/snowsql-1.2.5
git bisect good 03964e52b40738bd3066c78f2084ee5f9016c96e
# skip: [91104b5417275b780f6947b46a5c1bcc0d99f10b] Merge branch 'master' into staging-next
git bisect skip 91104b5417275b780f6947b46a5c1bcc0d99f10b
# bad: [566683ab5e76f570f34d0a4ec6484250859fae85] Merge staging into staging-next
git bisect bad 566683ab5e76f570f34d0a4ec6484250859fae85
# skip: [47d561b10fbe36be31c867151627cc0a955e0a63] libinput: 1.15.6 -> 1.16.1
git bisect skip 47d561b10fbe36be31c867151627cc0a955e0a63
# skip: [900b2d5a6c3fc150e7ef983fd7fb99a921585b22] python3: fix impure /bin/sh call in subprocess
git bisect skip 900b2d5a6c3fc150e7ef983fd7fb99a921585b22
# bad: [72c8ed038946850d345db13f30e259a368d7c3c5] systemd: build with cryptsetup and cryptsetup-generators
git bisect bad 72c8ed038946850d345db13f30e259a368d7c3c5
# first bad commit: [72c8ed038946850d345db13f30e259a368d7c3c5] systemd: build with cryptsetup and cryptsetup-generators
```

Combining this with https://github.com/NixOS/nixpkgs/issues/96479#issuecomment-682546223 I'm not really sure `nixos-generate-config` suddenly stops creating the 
```nix
boot.initrd.luks.devices."cryptroot".device = "...";
```

line.

Let's read some perl, I'd say :-)

**Comment by vcunat** [comment]
I confirm the bisection: the test fails on 72c8ed03 and succeeds on the parent commit.

**Comment by flokli** [comment]
Can we always add `dm_crypt` into initrd, until we understand why `nixos-generate-config` doesn't add that line? I assume reverting the commti will be a bit tough, as it resulted in a lot of cleanups (and will be a world-rebuild, too)

**Comment by jpdoyle** [comment]
https://github.com/NixOS/nixpkgs/blob/b843475bdd3087066eacc5e1c2d055afcd12c566/nixos/modules/installer/tools/nixos-generate-config.pl#L466 appears to be the snippet that generates that bit of config.

I'm not familiar with how that part of sysfs enumerates, but it seems like something that might be different when luks doesn't have udev

**Comment by jpdoyle** [comment]
From https://hydra.nixos.org/build/126109626/nixlog/2
```
machine: must succeed: echo -n supersecret | cryptsetup luksOpen --key-file - /dev/vda3 cryptroot
machine: must succeed: mkfs.ext3 -L nixos /dev/mapper/cryptroot
machine: [   27.947257] systemd-udevd[443]: dm-0: Conflicting device node '/dev/mapper/cryptroot' found, link to '/dev/dm-0' will not be created.
machine: mke2fs 1.45.5 (07-Jan-2020)
```

**Comment by flokli** [comment]
I did dig a bit further on the "systemd-udevd[443]: dm-0: Conflicting device node '/dev/mapper/cryptroot' log line from https://github.com/NixOS/nixpkgs/issues/96479#issuecomment-683217713. Thanks @jpdoyle for this!

Turns out, after the test runs the `cryptsetup luksOpen` command, *a block device file* with the same major and minor as `/dev/dm-0` is created in `/dev/mapper/cryptroot`.

Instead, `/dev/mapper/cryptroot` should be a *symlink* to `../dm-0` (according to `10-dm.rules` from the `lvm2` package, which is present in `/etc/udev/rules.d`.

The error message we saw is udev tripping over the following line in there:
> ENV{DM_UDEV_DISABLE_DM_RULES_FLAG}!="1", ENV{DM_NAME}=="?*", SYMLINK+="mapper/$env{DM_NAME}"

udev doesn't want to replace the device file with a symlink, and just keeps things as-is (which causes the warning in dmesg).

I tried manually changing `/dev/mapper/cryptroot` back to a symlink, which got `nixos-generate-config` to properly detect things, `dm_crypt` being added to initrd, and `boot-after-install` to succeed booting:

```
From 235294b8b60ebd0233a3a10e01ce32e9f4e4527a Mon Sep 17 00:00:00 2001
From: Florian Klink <flokli@flokli.de>
Date: Sat, 29 Aug 2020 19:45:50 +0200
Subject: [PATCH] nixosTests.installer: workaround /dev/mapper/cryptroot not
 being a symlink to /dev/dm-0

For some reason, cryptsetup luksOpen currently causes a device node to
be created in /dev/mapper/cryptroot, not a symlink to /dev/dm-0, which
causes nixos-generate-config to not detect the crypt device (and by
this, not shipping dm_crypt inside initrd, failing boot).

See https://github.com/NixOS/nixpkgs/issues/96479
---
 nixos/tests/installer.nix | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/nixos/tests/installer.nix b/nixos/tests/installer.nix
index f64e4101d45..ec5369d858e 100644
--- a/nixos/tests/installer.nix
+++ b/nixos/tests/installer.nix
@@ -358,6 +358,12 @@ let
             "modprobe dm_mod dm_crypt",
             "echo -n supersecret | cryptsetup luksFormat ${luksFormatOpts} -q /dev/vda3 -",
             "echo -n supersecret | cryptsetup luksOpen --key-file - /dev/vda3 cryptroot",
+            # TODO: for some reason, above command causes a device node to be
+            # created in /dev/mapper/cryptroot, not a symlink to /dev/dm-0,
+            # which causes nixos-generate-config to not detect the crypt device
+            # (and by this, not shipping dm_crypt inside initrd, failing boot)
+            # see https://github.com/NixOS/nixpkgs/issues/96479
+            "ln -sfn ../dm-0 /dev/mapper/cryptroot",
             "mkfs.ext3 -L nixos /dev/mapper/cryptroot",
             "mount LABEL=nixos /mnt",
             "mkfs.ext3 -L boot /dev/vda1",
-- 
2.28.0

```
I still don't fully understand
 - how the device node is created (is it the `cryptsetup` binary, or some library code it's linked against (libdevmapper from lvm2), or another udev rule we don't want to have, or something else?)
 - if this is intended behaviour. I will check a bit how things look on NixOS systems before the systemd-cryptsetup commit, as well as other distros.
 - If we can improve `nixos-generate-config` to still detect things properly if these are just device nodes, and not symlinks.

i propose applying above patch to `master` and `staging(-next)` to unblock channels, but we need to better understand and apply more fixes until the 20.09 release - `nixos-generate-config` might not create a bootable `hardware-configuration.nix` currently.

**Comment by arianvp** [comment]
`/dev/mapper/cryptroot` being a _device_ instead of a symlink is very confusing. No idea why that would be happening

From https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/logical_volume_manager_administration/udev_device_manager I deduce that `libdevmapper` can fall back to creating device nodes itself when `udev` support is disabled (which you do by setting `udev = null`) but that's only in `systemd`; not in `cryptsetup` package in general... so bit confused

**Comment by arianvp** [comment]
A little reproducer:
```
truncate -s 1G lol
echo -n lol | cryptsetup luksFormat  lol
echo -n lol | sudo cryptsetup luksOpen --key-file - lol lol
ls -la /dev/mapper/lol
sudo cryptsetup close lol
```

On NixOS 20.03 this returns:

```
lrwxrwxrwx 1 root root 7 Aug 29 21:28 /dev/mapper/lol -> ../dm-1
```

a symlink as expected.

**Comment by emmanuelrosa** [comment]
On my NixOS machine I get this:

```
Aug 30 07:29:54 nixos-laptop stage-1-init: Passphrase for /dev/disk/by-uuid/***:
Aug 30 07:29:54 nixos-laptop stage-1-init: Verifying passphrase for /dev/disk/by-uuid/***... - success
Aug 30 07:29:54 nixos-laptop stage-1-init: Passphrase for /dev/disk/by-uuid/***: reused
Aug 30 07:29:54 nixos-laptop kernel: random: cryptsetup: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[141]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop stage-1-init: Verifying passphrase for /dev/disk/by-uuid/***... - success
Aug 30 07:29:54 nixos-laptop stage-1-init: starting device mapper and LVM...
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[180]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
Aug 30 07:29:54 nixos-laptop kernel: BTRFS: device label NIXOS devid 2 transid 407719 /dev/mapper/luksrootfs2
Aug 30 07:29:54 nixos-laptop kernel: BTRFS: device label NIXOS devid 1 transid 407719 /dev/mapper/luksrootfs1
Aug 30 07:29:54 nixos-laptop stage-1-init: Scanning for Btrfs filesystems
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[124]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[133]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop systemd-udevd[178]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[135]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[180]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop systemd-udevd[124]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[141]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[180]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
Aug 30 07:29:54 nixos-laptop systemd-udevd[178]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[133]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[178]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop systemd-udevd[125]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop systemd-udevd[129]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[129]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop systemd-udevd[180]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
Aug 30 07:29:54 nixos-laptop kernel: random: lvm: uninitialized urandom read (4 bytes read)
Aug 30 07:29:54 nixos-laptop kernel: random: crng init done
Aug 30 07:29:54 nixos-laptop systemd-udevd[140]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[123]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
Aug 30 07:29:54 nixos-laptop systemd-udevd[128]: dm-0: Conflicting device node '/dev/mapper/luksrootfs1' found, link to '/dev/dm-0' will not be created.
Aug 30 07:29:54 nixos-laptop systemd-udevd[138]: dm-1: Conflicting device node '/dev/mapper/luksrootfs2' found, link to '/dev/dm-1' will not be created.
```

After all the LUKS containers are unlocked the system continues to boot as normal.

**Comment by doronbehar** [comment]
>     There's a circular dependency to systemd via cryptsetup and lvm2
>     (systemd -> cryptsetup -> lvm2 -> udev=systemd).

There are many other circular dependencies in NixOS (known issue for all distros, see [discussion](https://discourse.nixos.org/t/guix-reduces-bootstrap-seed-by-50/4304)). Perhaps would it be possible to address it the same way we address `autoconf` dependency in `autoconf` and alike?

**Comment by vcunat** [comment]
I don't follow.  So far it doesn't seem that the circularity is really related to the problem we have here.  And I have no idea what you mean by "address it the same way".

**Comment by doronbehar** [comment]
> I don't follow. So far it doesn't seem that the circularity is really related to the problem we have here.

It's likely that you shouldn't listen to me. But, I got the slightest impression from the bisect that if lvm2 is not built with udev, it doesn't detect the encrypted devices as before.

>  And I have no idea what you mean by "address it the same way".

The idea I had in mind was no good, now I realise - we can't get a binary release of systemd and bootstrap it from there.

**Comment by andersk** [comment]
I found that taking one more step along the systemd → cryptsetup → lvm2 → udev=systemd dependency cycle before breaking the chain allows the test to pass.

```diff
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -17237,7 +17237,7 @@ in
   cryptsetup = callPackage ../os-specific/linux/cryptsetup {
     # cryptsetup only really needs the devmapper component of cryptsetup
     # but itself is used as a library in systemd (=udev)
-    lvm2 = lvm2.override { udev = null; };
+    lvm2 = lvm2.override { udev = udev.override { cryptsetup = null; }; };
   };
 
   cramfsprogs = callPackage ../os-specific/linux/cramfsprogs { };
```

#97002

**Comment by doronbehar** [comment]
> This is a bit ridiculous because we end up with this even stupider chain in the runtime closure of lvm2:
> 
> lvm2 → systemd → cryptsetup → different lvm2 → different systemd

Wow, amazing. Do other Linux distros don't care? because they don't distinct between a package's inputs and the package itself? For sure Guix hasn't encountered this issue since they don't even use systemd! See [this](https://lists.gnu.org/archive/html/guix-devel/2018-04/msg00001.html) and [this](https://lists.gnu.org/archive/html/guix-devel/2018-04/msg00050.html).

**Comment by arianvp** [comment]
Other distro's have been running into the exact same issue:

https://www.openembedded.org/pipermail/openembedded-devel/2018-February/116976.html
https://patchwork.openembedded.org/patch/139216/
https://github.com/distr1/distri/issues/16

**Comment by vcunat** [comment]
For reference, any cycle-breaking should also watch the issue addressed by f6286dea88, as that solution isn't really "foolproof", i.e. such changes might again lead to breakages that happen based on non-determinstic issues (based on filesystem ordering).

**Comment by andersk** [comment]
@vcunat I believe I’ve now addressed that issue by simply reverting f6286dea8881e446f33a9b2faca8db43ec55aa45; see my [explanation](https://github.com/NixOS/nixpkgs/pull/97008#issuecomment-686691514_) in #97008.

**Comment by flokli** [comment]
https://github.com/NixOS/nixpkgs/pull/97008 has been merged, cc @FRidh so it can trickle through the next staging round.

**Comment by arianvp** [comment]
Lets close this; and open a new issue on breaking the cycle in a less closure-size-bloating way (e.g. by separating libudev)

**Comment by Atemu** [comment]
Why do we need to wait for staging? Is master even in a usable state without the patch? Is it that staging will be merged very soon anyways?

**Comment by flokli** [comment]
master works, except installations from a master-built install medium will fail to detect crypto volumes (which is why the installer tests fail and the channel doesn't advance).

This will be backported into 20.09 before the release one way or another - we just don't want to merge it without hydra having built most of the packages - otherwise ZHF will get pretty annoying.

**Comment by FRidh** [comment]
[Staging next](https://github.com/NixOS/nixpkgs/pull/97146) that has this change. Evaluation is ongoing. If tomorrow morning or around noon things look good it can be merged into master (and I suppose we branch off then).

**Comment by jonringer** [comment]
> [Staging next](https://github.com/NixOS/nixpkgs/pull/97146) that has this change. Evaluation is ongoing. If tomorrow morning or around noon things look good it can be merged into master (and I suppose we branch off then).

This will likely be the case. However, I'm traveling

**Comment by vcunat** [comment]
It's rebuilding stdenv ATM, so I don't think we'll have that many binaries around the noon yet.

**Comment by vcunat** [comment]
The test should work in current master.
